src/check/parse/tokenize/src/main.zig

```diff
@@ -0,0 +1,1399 @@
+                const info = self.peekUnicode();
+                if (info.kind != UnicodeKind.Other and info.kind != UnicodeKind.Invalid) {
+                    self.pos += info.length;
+                } else {
+                    break;
+                }
+            }
+        }
+    }
+};
+
+/// The output of the tokenizer.
+pub const TokenOutput = struct {
+    tokens: TokenizedBuffer,
+    /// The messages slice is the same one provided by the caller.
+    messages: []Message,
+    message_count: usize,
+};
+
+const BraceKind = enum {
+    Round,
+    Square,
+    Curly,
+    StringInterpolation,
+    StringInterpolationMultiline,
+    SingleQuoteInterpolation,
+    SingleQuoteInterpolationMultiline,
+};
+
+/// The tokenizer that uses a Cursor and produces a TokenizedBuffer.
+pub const Tokenizer = struct {
+    cursor: Cursor,
+    output: TokenizedBuffer,
+    stack: std.ArrayList(BraceKind),
+
+    /// Creates a new Tokenizer.
+    /// Note that the caller must also provide a pre-allocated messages buffer.
+    pub fn init(text: []const u8, messages: []Message, gc: *GenCatData, allocator: std.mem.Allocator) !Tokenizer {
+        const cursor = Cursor.init(text, messages, gc);
+        var output = try TokenizedBuffer.init(allocator);
+        // Push an initial line with indent 0.
+        try output.lines.append(Line{
+            .token_start = 0,
+            .indent = Indent.init(),
+        });
+        return Tokenizer{
+            .cursor = cursor,
+            .output = output,
+            .stack = std.ArrayList(BraceKind).init(allocator),
+        };
+    }
+
+    pub fn destroy(self: *Tokenizer) void {
+        // self.output.kinds.deinit();
+        // self.output.offsets.deinit();
+        // self.output.lengths.deinit();
+        // self.output.lines.deinit();
+        self.stack.deinit();
+        self.gc.deinit();
+    }
+
+    // A simplified equivalent to the Rust macros:
+    fn pushToken(self: *Tokenizer, kind: T, start: usize) !void {
+        const len = self.cursor.pos - start;
+        try self.output.pushToken(kind, start, len);
+    }
+
+    fn consumeBraceCloseAndContinueStringInterp(self: *Tokenizer, brace: BraceKind) !void {
+        std.debug.assert(self.cursor.peek() == '}' or self.cursor.peek() == ']' or self.cursor.peek() == ')');
+        if (self.stack.items.len == 0) {
+            self.cursor.pushMessageHere(.OverClosedBrace);
+            self.cursor.pos += 1;
+            return;
+        }
+        const last = self.stack.items[self.stack.items.len - 1];
+        if (last == brace) {
+            self.stack.items = self.stack.items[0 .. self.stack.items.len - 1];
+        } else {
+            self.cursor.pushMessageHere(.MismatchedBrace);
+        }
+        self.cursor.pos += 1;
+        const start = self.cursor.pos;
+        if (self.stack.items.len > 0) {
+            switch (self.stack.items[self.stack.items.len - 1]) {
+                BraceKind.StringInterpolation => {
```

Comment:
Note: I think you can just do `.StringInterpolation` here and drop the `BraceKind`